### DISCUSSION TITLE
How can I call a PyObject from an animation frame callback using its heap address?

### DISCUSSION THREAD

**Discussion by szabolcsdombi** [pyodide/pyodide#3714] [Q&A]
Hi.

I found `pyodide._module.callPyObject(ptr, []);` and it works perfectly for my use case except that sometimes It throws an exception:

![Screenshot_72](https://user-images.githubusercontent.com/11232402/228783529-3e5a52aa-4300-40a9-a961-975573cbf85d.png)

I have checked and the python refcounts are fine.

This happens very rarely, and interestingly only after I added the event handlers for the canvas that has nothing to do with the wasm runtime module. It also happens more often with scenes that load a lot of mesh (maybe heap resize related as well).

I am probably off the right track.
So my questions are:

- how can I do this with the documented API?
- Is it safe to call it from within the callback of `requestAnimationFrame`.

Please note that I have only a heap pointer to the Python object I cannot pass JS objects to this function.

Here is a [live demo](https://szabolcsdombi.github.io/zengl-webgl-demo/food.html).

This line causes the stack trace: [code](https://github.com/szabolcsdombi/zengl/blob/eb5042200c0e8ed32a62e4d52cbab80274ffcd57/examples/webgl/public/webgl.js#L95)

It is called from a Python Extension module: [code](https://github.com/szabolcsdombi/zengl/blob/eb5042200c0e8ed32a62e4d52cbab80274ffcd57/examples/webgl/webgl/webgl.c#L666-L667)

PS: here is a [demo](https://szabolcsdombi.github.io/zengl-webgl-demo/multi_canvas_support.html) for having 6 of these callbacks. this almost never fails due to the simplicity of the scenes rendered.

**Comment by hoodmane** [comment]
I need a bit more explanation of why you only have the pointer. Presumably the issue is that you're not holding a refcount on the PyObject so calling it is unsound. But why not just call a PyProxy like normal?

**Comment by szabolcsdombi** [comment]
I want to achieve the following:

When the Python script executes (for a game, simulation, ...) the initialization part is executed and the per frame update/render is not.

Here is a simple example how it works right now. [code](https://github.com/szabolcsdombi/zengl/blob/eb5042200c0e8ed32a62e4d52cbab80274ffcd57/examples/webgl/public/triangle.html#L34-L109)

The non-web examples have the main loop in the control of the programmer like here. [example](https://github.com/szabolcsdombi/zengl/blob/eb5042200c0e8ed32a62e4d52cbab80274ffcd57/examples/hello_triangle.py#L50-L55)

Since this is not possible in webassembly, the execution must terminate in order for the UI to be responsive. I delegated the main loop partially to a callback that is set by the `@window.render` to a python function.

This way the examples remain simple and using classes is not a must. not a must but possible, in that case the `window.render` is use like an ordinary function to set the callback. [code](https://github.com/szabolcsdombi/zengl/blob/eb5042200c0e8ed32a62e4d52cbab80274ffcd57/examples/webgl/public/multi_canvas_support.html#L51)

When I create the window (canvas) from Python, I want to install the callback to a function that does the following:

- It Updates the Window class (right now [this](https://github.com/szabolcsdombi/zengl/blob/eb5042200c0e8ed32a62e4d52cbab80274ffcd57/examples/webgl/webgl/webgl.c#L678-L682) part)
- Then it calls the render callback function set by the user (if any).

The window update part is responsible for reading the user input from JS and set some window properies.

When I create the window (canvas) the call is made from the C++ code and handled by JS, I cannot pass anything else but scalars and pointers to it.

**Reply by hoodmane** [reply]
> the execution must terminate in order for the UI to be responsive

Have you tried something like:
```py
from asyncio import sleep, ensure_future
async def main_loop():
	while window.update():
	    ctx.new_frame()
		await sleep(0)
	    image.clear()
        await sleep(0)
	    pipeline.render()
        await sleep(0)
	    image.blit()
        await sleep(0)
	    ctx.end_frame()
        await sleep(0)

ensure_future(main_loop())
```

**Reply by hoodmane** [reply]
What about putting Python into a webworker?

**Reply by hoodmane** [reply]
I haven't read what you said super carefully yet, just providing some generic suggestions.

**Reply by hoodmane** [reply]
>  the call is made from the C++ code and handled by JS

Okay this is an interesting issue. You need the Pyodide C API. Which isn't well documented or stable or anything. I'll give an example of what this might look like for now when I get a moment.
